feat: Added Video inference feature into unsloth_zoo/vision_utils.py from qwen-vl-utils

autinn opened the pull request:
**Problem**

Currently, Unsloth's vision pipeline is designed primarily for static image inputs. It lacks the native functionality to process video files directly within the `unsloth_zoo` library. This prevents users from performing video-based inference with a pre-trained `Qwen2.5-VL` model without relying on external, non-integrated dependencies (e.g. `qwen-vl-utils`) .

This also points to the feature requests for processing video inputs ([#3061](https://github.com/unslothai/unsloth/issues/3016)). 

**Solution** 

Our team, (@neenz16, @suchithgali, and @autinn), created this pull request to introduces a complete, self-contained solution for video inference within the `unsloth_zoo` library. Instead of adding an external dependency from `Qwen2.5-VL/qwen-vl-utils`, we implement the core video processing logic directly in `unsloth_zoo/vision_utils.py` using the source code from `qwen_vl_utils/vision_process.py`. 

**Key changes include:**

- **Implemented a Full Video Pipeline**: The necessary functions to read, decode, and sample frames from video files, including `smart_nframes`, `_read_video_torchvision`, `_read_video_decord`, `_read_video_torchcodec`, and their helpers, have been added. This makes the unsloth_zoo library self-contained for video processing.

- **Dynamic Video Sampling**: The `fetch_video` function has been fully implemented to handle video files, intelligently sampling frames, and managing resolution and other model constraints.

- **Processor Integration**: The `process_vision_info` function and `UnslothVisionDataCollator` are updated to correctly receive and pass the video tensor and its associated metadata (like FPS) to the model's processor.

- **Debug Print Statement**: Added a print statement to indicate and verify `unsloth_zoo/vision_utils` is running for video reading with the specified backend. 

**Test** 

This provided [Colab Notebook](https://colab.research.google.com/drive/1VoIvSsXyY_TXgP2lflUl1i_FdhAIweNT?usp=sharing) demonstrates the before and after for running an end-to-end video inference run with and without the `qwen-vl-utils` dependency.

**Before (Missing Dependency on Video Processing)**: A test shows that without installing `qwen-vl-utils`, a user attempting to run the original Qwen video inference code would encounter a ModuleNotFoundError.

**Before (With External Dependency on qwen-vl-utils)**: The notebook shows that when` qwen-vl-utils` is installed, a user can successfully run video inference using the library's functions. This establishes a baseline for the desired functionality.

**After (Integrated qwen-vl-utils to unsloth)**: The new test script successfully loads the `unsloth/Qwen2.5-VL-7B-Instruct-bnb-4bit` model. It then overwrites the `vision_utils.py` in `unsloth_zoo` with the newly implemented code to process a video file and a textual prompt. The test runs a single inference call with `model.generate()` and confirms that the model successfully processes the video and generates a coherent text response, all without relying on the external dependency.

**Documentation Suggestion**

We also propose a new “Video Inference” section on the [“Vision Fine-tuning”](https://docs.unsloth.ai/basics/vision-fine-tuning) part of the Unsloth documentation as the first step to achieve video fine-tuning. We are happy to work on that as well after the PR is merged to main.

**Reference**
qwen-vl-utils documentation: https://github.com/QwenLM/Qwen2.5-VL/tree/main/qwen-vl-utils 
qwen vision_process.py code: https://github.com/QwenLM/Qwen2.5-VL/blob/main/qwen-vl-utils/src/qwen_vl_utils/vision_process.py

mmathew23 commented on unsloth_zoo/vision_utils.py:
```diff
@@ -54,21 +54,39 @@
 VIDEO_MAX_PIXELS = 768 * 28 * 28
-VIDEO_TOTAL_PIXELS = 24576 * 28 * 28
+VIDEO_TOTAL_PIXELS = int(float(os.environ.get('VIDEO_MAX_PIXELS', 128000 * 28 * 28 * 0.9)))
+logger.info(f"set VIDEO_TOTAL_PIXELS: {VIDEO_TOTAL_PIXELS}")
```
Using the logger is fine, but we generally want to keep logging less cluttered unless the user wants it. In other parts of the library we check for `do_logging = os.environ.get("UNSLOTH_ENABLE_LOGGING", "0") == "1"` to determine whether or not to log. If you'd like to stick with logger I'd suggest configuring the logging level based on the env variable, and no prints/logs if the variable is off, and prints if the variable is on. Best to also prepend "Unsloth: " before the actual string to log.

mmathew23 commented on unsloth_zoo/vision_utils.py:
```diff
@@ -172,6 +190,310 @@ def fetch_image(
+        max_frames = floor_by_factor(ele.get("max_frames", min(FPS_MAX_FRAMES, total_frames)), FRAME_FACTOR)
+        nframes = total_frames / video_fps * fps
+        if nframes > total_frames:
+            logger.warning(f"smart_nframes: nframes[{nframes}] > total_frames[{total_frames}]")
```
check if logging is enabled before logging.

mmathew23 commented on unsloth_zoo/vision_utils.py:
```diff
@@ -172,6 +190,310 @@ def fetch_image(
+        output_format="TCHW",
+    )
+    total_frames, video_fps = video.size(0), info["video_fps"]
+    logger.info(f"torchvision:  {video_path=}, {total_frames=}, {video_fps=}, time={time.time() - st:.3f}s")
```
check if logging is enabled before logging and include Unsloth:

mmathew23 commented on unsloth_zoo/vision_utils.py:
```diff
@@ -172,6 +190,310 @@ def fetch_image(
+    idx = torch.linspace(start_frame, end_frame, nframes).round().long().tolist()
+    video = vr.get_batch(idx).asnumpy()
+    video = torch.tensor(video).permute(0, 3, 1, 2)  # Convert to TCHW format
+    logger.info(f"decord:  {video_path=}, {total_frames=}, {video_fps=}, time={time.time() - st:.3f}s")
```
check if logging is enabled before logging and include Unsloth:

mmathew23 commented on unsloth_zoo/vision_utils.py:
```diff
@@ -172,6 +190,310 @@ def fetch_image(
+        try:
+            video, sample_fps = VIDEO_READER_BACKENDS[video_reader_backend](ele)
+        except Exception as e:
+            logger.warning(f"video_reader_backend {video_reader_backend} error, use torchvision as default, msg: {e}")
```
check if logging is enabled before logging and include Unsloth:

neenz16 commented:
> Hi @autinn I added some comments. Since there's also changes to the collator we should make sure that existing training runs work as intended, while video models can also train too.
> 
> Great work, and appreciate the help! Once we get the comments and finetuning verification we can merge this.

Thanks a lot for the review and guidance @mmathew23! 🙏
We’ll work on the suggested changes and run verifications to ensure both existing training runs and video model finetuning work as expected. We’ll get back with updates soon. Really appreciate your support!

autinn commented:
@mmathew23 Thank you again for your feedback! We worked on the code for a bit the past few days and here are the changes we made: 

1. Remove `from __future__ import annotations` 
2. Import logger from `.log` and `UNSLOTH_ENABLE_LOGGING` from `.temporary_patches.common`, similar to vllm_utils.py
3. Added conditional logging for the `logger.info()` and `logger.warning()` statements, including the print statement we originally added.
4. Updated variable naming from `FORCE_QWENVL_VIDEO_READER` to `FORCE_UNSLOTH_VIDEO_READER`
5. Tested on Colab Notebook and saw all logging statements showed as intended.

mmathew23 commented:
Hello @autinn @neenz16 @suchithgali! Appreciate all the help on this.

We had refactored the vision_utils code before getting this in. I also had to clean up some issues to make sure non video finetuning still worked the same as well. I didn't have access to edit this PR branch directly so I created a new PR with your contributions, merged it into the new code, and cleaned it up. You can check it out here, https://github.com/unslothai/unsloth-zoo/pull/279. Thank you so much for your contribution!

neenz16 commented:
Thanks so much, @mmathew23! Really appreciate you merging and cleaning this up. Glad our contributions made it in — We'll check out #279!
